src/remoted/manager.c

```diff
@@ -277,8 +277,60 @@ void save_controlmsg(const keyentry * key, char *r_msg, size_t msg_length, int *
         default:
             break;
         }
-        if (strcmp(r_msg, HC_STARTUP) == 0) {
+        if (strncmp(r_msg, HC_STARTUP, strlen(HC_STARTUP)) == 0) {
             mdebug1("Agent %s sent HC_STARTUP from '%s'", key->name, aux_ip);
+
+            cJSON *agent_info = NULL;
+            cJSON *version = NULL;
+            if (agent_info = cJSON_Parse(strchr(r_msg, '{')), agent_info) {
+                char *error_msg_string = NULL;
+                if (version = cJSON_GetObjectItem(agent_info, "version"), cJSON_IsString(version)) {
+                    if (compare_wazuh_versions(__ossec_version, version->valuestring, false) < 0) {
+                        /* Reply to the agent */
+                        cJSON *error_msg = cJSON_CreateObject();
+                        cJSON_AddStringToObject(error_msg, "message", HC_INVALID_VERSION);
+                        error_msg_string = cJSON_PrintUnformatted(error_msg);
+                        char msg_err[OS_FLSIZE + 1] = "";
+                        snprintf(msg_err, OS_FLSIZE, "%s%s%s", CONTROL_HEADER, HC_ERROR, error_msg_string);
+                        send_msg(key->id, msg_err, -1);
+                        mdebug2("Unable to connect agent: %s. %s", key->id, HC_INVALID_VERSION);
+                        result = wdb_update_agent_status_code(atoi(key->id), INVALID_VERSION, version->valuestring, logr.worker_node ? "syncreq" : "synced", wdb_sock);
+
+                        if (OS_SUCCESS != result) {
+                            mwarn("Unable to set status code for agent: %s", key->id);
+                        }
+
+                        cJSON_Delete(agent_info);
+                        cJSON_Delete(error_msg);
+                        os_free(error_msg_string);
+                        os_free(clean);
+
+                        return;
+                    }
+                } else {
+                    merror("Error getting version from agent '%s'", key->id);
+
+                    cJSON *error_msg = cJSON_CreateObject();
```

Comment:
This code is almost the same than above. Can you create a function to avoid duplicating it?